### PR TITLE
[PDI-18369] Converting Integer to Timestamp type using Select values …

### PR DESCRIPTION
**Pull request by smmribeiro**
…step results in incorrect date

The Kettle property is now called 'KETTLE_TIMESTAMP_NUMBER_CONVERSION_MODE' and controls how Timestamp should be converted to a number **and vice-versa**.
There're three possible values:
- 'LEGACY': (the default, the behaviour that existed until now) Timestamp to number uses milliseconds but number to Timestamp uses nanoseconds
- 'MILLISECONDS': both conversions use milliseconds
- 'NANOSECONDS': both conversions use nanoseconds

Added several unit tests to guarantee all three scenarios (and a few others to upper the coverage).

@pentaho-lmartins @ssamora @ppatricio @bcostahitachivantara

### CODE REVIEW

**Comment by buildguy** [comment]
### :warning: Build finished in 34m 29s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 18 test(s) failed:

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToNumber_Milliseconds</b> (click to expand)</summary><p>

```
expected:<1.567308896123E12> but was:<1.567312496123E12>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToBigNumber_Nanoseconds</b> (click to expand)</summary><p>

```
expected:<1567308896123456789> but was:<1567312496123456789>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToNumber_Nanoseconds</b> (click to expand)</summary><p>

```
expected:<1.56730889612345677E18> but was:<1.56731249612345677E18>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToBigNumber_DefaultMode</b> (click to expand)</summary><p>

```
expected:<1567308896123> but was:<1567312496123>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToInteger_Milliseconds</b> (click to expand)</summary><p>

```
expected:<1567308896123> but was:<1567312496123>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertBigNumberToTimestamp_Milliseconds</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123> but was:<2019-09-01 04:34:56.123>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertIntegerToTimestamp_Milliseconds</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123> but was:<2019-09-01 04:34:56.123>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToNumber_DefaultMode</b> (click to expand)</summary><p>

```
expected:<1.567308896123E12> but was:<1.567312496123E12>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToInteger_Nanoseconds</b> (click to expand)</summary><p>

```
expected:<1567308896123456789> but was:<1567312496123456789>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertIntegerToTimestamp_Nanoseconds</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123456789> but was:<2019-09-01 04:34:56.123456789>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToInteger_DefaultMode</b> (click to expand)</summary><p>

```
expected:<1567308896123> but was:<1567312496123>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertTimestampToBigNumber_Milliseconds</b> (click to expand)</summary><p>

```
expected:<1567308896123> but was:<1567312496123>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertIntegerToTimestamp_DefaultMode</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123456789> but was:<2019-09-01 04:34:56.123456789>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertNumberToTimestamp_Nanoseconds</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123456> but was:<2019-09-01 04:34:56.123456>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertBigNumberToTimestamp_Nanoseconds</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123456789> but was:<2019-09-01 04:34:56.123456789>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertNumberToTimestamp_Milliseconds</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123> but was:<2019-09-01 04:34:56.123>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertNumberToTimestamp_DefaultMode</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123456> but was:<2019-09-01 04:34:56.123456>
```

</p></details>

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaTimestampTest.testConvertBigNumberToTimestamp_DefaultMode</b> (click to expand)</summary><p>

```
expected:<2019-09-01 05:34:56.123456789> but was:<2019-09-01 04:34:56.123456789>
```

</p></details>

  
Tests run: **4789**, Failures: **18**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7364/1/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 36m 5s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine
```
---
:ok_hand: All tests passed!
 
Tests run: **4789**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7364/2/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by smmribeiro** [comment]
Please, don't forget to squash commits!

**Comment by buildguy** [comment]
### :warning: Build finished in 38m 2s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 1 test(s) failed:

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaConverterTest.testTimestampIntegerConversions</b> (click to expand)</summary><p>

```
expected:<2019-09-01 04:34:56.123456789> but was:<2019-09-01 03:34:56.123456789>
```

</p></details>

  
Tests run: **4790**, Failures: **1**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7364/3/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :warning: Build finished in 38m 30s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine
```
---
### :no_entry: Failed Tests

#### :cloud_with_lightning_and_rain: 1 test(s) failed:

<details><summary><b>org.pentaho.di.core.row.value.ValueMetaConverterTest.testTimestampIntegerConversions</b> (click to expand)</summary><p>

```
expected:<2019-09-01 04:34:56.123456789> but was:<2019-09-01 03:34:56.123456789>
```

</p></details>

  
Tests run: **4790**, Failures: **1**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7364/4/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 1h 23m 18s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine
```
---
:ok_hand: All tests passed!
 
Tests run: **4790**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7364/5/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

File: core/src/main/java/org/pentaho/di/core/row/value/ValueMetaTimestamp.java [hunk 1]
```diff
@@ -348,14 +341,14 @@ public Object convertDataFromString( String pol, ValueMetaInterface convertMeta,
...
       return null;
     } else {
-      // if the null_value is specified, we try to match with that.
+      // if the nullValue is specified, we try to match with that.
       //
-      if ( !Utils.isEmpty( null_value ) ) {
-        if ( null_value.length() <= pol.length() ) {
-          // If the polled value is equal to the spaces right-padded null_value,
+      if ( !Utils.isEmpty( nullValue ) ) {
+        if ( nullValue.length() <= pol.length() ) {
```
Comment (peterrinehart) [inline core/src/main/java/org/pentaho/di/core/row/value/ValueMetaTimestamp.java]: Sonar is right; these two IF statements can be merged together.

**Comment by buildguy** [comment]
### :white_check_mark: Build finished in 38m 1s
Build command:
```console
mvn clean verify -B -e -Daudit -amd -pl core,engine
```
---
:ok_hand: All tests passed!
 
Tests run: **4790**, Failures: **0**, Skipped: **7** &nbsp;&nbsp; [Test Results](http://wingman2.pentaho.net/job/projects/job/pdi/job/PR-7364/6/testReport)

 ---
 
 

:information_source:  _This is an automatic message_

**Pull request merged**